fft/src/KokkosFFT_FFTW_Types.hpp

```diff
@@ -0,0 +1,158 @@
+  static constexpr FFTWTransformType type() { return m_type; };
+};
+
+template <typename T1, typename T2>
+struct fftw_transform_type<Kokkos::complex<T1>, Kokkos::complex<T2>> {
+  static_assert(std::is_same_v<T1, T2>,
+                "T1 and T2 should have the same precision");
+  static constexpr FFTWTransformType m_type = std::is_same_v<T1, float>
+                                                  ? FFTWTransformType::C2C
+                                                  : FFTWTransformType::Z2Z;
+  static constexpr FFTWTransformType type() { return m_type; };
+};
+
+/// \brief A class that wraps fftw_plan and fftwf_plan for RAII
+template <typename ExecutionSpace, typename T1, typename T2>
+struct ScopedFFTWPlan {
+ private:
+  using floating_point_type = KokkosFFT::Impl::base_floating_point_type<T1>;
+  using plan_type =
+      std::conditional_t<std::is_same_v<floating_point_type, float>, fftwf_plan,
+                         fftw_plan>;
+  plan_type m_plan;
+
+ public:
+  template <typename InScalarType, typename OutScalarType>
+  ScopedFFTWPlan(const ExecutionSpace &exec_space, int rank, const int *n,
+                 int howmany, InScalarType *in, const int *inembed, int istride,
+                 int idist, OutScalarType *out, const int *onembed, int ostride,
+                 int odist, [[maybe_unused]] int sign, unsigned flags) {
+    init_threads(exec_space);
+    constexpr auto type = fftw_transform_type<T1, T2>::type();
+    if constexpr (type == KokkosFFT::Impl::FFTWTransformType::R2C) {
+      m_plan =
+          fftwf_plan_many_dft_r2c(rank, n, howmany, in, inembed, istride, idist,
+                                  out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::D2Z) {
+      m_plan =
+          fftw_plan_many_dft_r2c(rank, n, howmany, in, inembed, istride, idist,
+                                 out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::C2R) {
+      m_plan =
+          fftwf_plan_many_dft_c2r(rank, n, howmany, in, inembed, istride, idist,
+                                  out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::Z2D) {
+      m_plan =
+          fftw_plan_many_dft_c2r(rank, n, howmany, in, inembed, istride, idist,
+                                 out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::C2C) {
+      m_plan =
+          fftwf_plan_many_dft(rank, n, howmany, in, inembed, istride, idist,
+                              out, onembed, ostride, odist, sign, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::Z2Z) {
+      m_plan = fftw_plan_many_dft(rank, n, howmany, in, inembed, istride, idist,
+                                  out, onembed, ostride, odist, sign, flags);
+    }
+  }
+
+  ~ScopedFFTWPlan() noexcept {
+    cleanup_threads();
```

Comment:
It looks better but I am afraid it easily leads to UB. If we take the following scenario:
```cpp
KokkosFFt::Plan plan( /* ... */ );
KokkosFFt::fft( /* ... */ );
plan.execute( /* ... */ );
``` 
My understanding is that the call to `KokkosFFt::fft` will cleanup threads in the destructor of the internal `KokkosFFT::Plan`. The execution of the second FFT transform with `plan.execute` is done after a call to `fftw_cleanup_threads`, which is not allowed by FFTW https://www.fftw.org/doc/Usage-of-Multi_002dthreaded-FFTW.html

Reply:
I tried to fix that with an unique id of `ScopedFFTWPlan`.

Reply:
I found the current implementation is sill incorrect. We need to call `fftw_cleanup_threads` in the destructor of the final instance of `ScopedFFTWPlan`. However, there is no way to identify the final instance. 

For the moment, we need to suppress `cleanup_threads()`.
We may have two options. 

1. Introduce `KokkosFFT::initialize` and `KokkosFFT::finalize` to wrap global initialization and cleanup
2. Parallelize fftw exec manually by `Kokkos`.

Reply:
> I found the current implementation is sill incorrect. We need to call `fftw_cleanup_threads` in the destructor of the final instance of `ScopedFFTWPlan`. However, there is no way to identify the final instance.
> 
> For the moment, we need to suppress `cleanup_threads()`. We may have two options.
> 
> 1. Introduce `KokkosFFT::initialize` and `KokkosFFT::finalize` to wrap global initialization and cleanup
> 2. Parallelize fftw exec manually by `Kokkos`.

I think I have a slight preference for the solution 1. as it will also help to solve #217. What do you think ?

Reply:
I may have another solution with a [local static object](https://godbolt.org/z/EMbYa5Ta7)

Reply:
How about removing `fftw_cleanup_threads` from this PR and make it by solution 1 in another PR?

Reply:
Yes looks a good idea

Reply:
@tpadioleau May I get another review. Note it is safe to call `fftw_init_threads` multiplie times.